syntax = "proto3";

package search;

import "google/protobuf/empty.proto";

service Index {
    rpc takeNext (google.protobuf.Empty) returns (TakeNextResponse);
    rpc putNew (PutNewRequest) returns (google.protobuf.Empty);
    rpc addToIndexWords (AddToIndexRequestWords) returns (google.protobuf.Empty);
    rpc addToIndexLinks (AddToIndexRequestLinks) returns (google.protobuf.Empty);
    rpc searchWord (SearchWordRequest) returns (SearchWordResponse);
    rpc searchBacklinks (SearchBacklinksRequest) returns (SearchBacklinksResponse);
    rpc getFullIndex (google.protobuf.Empty) returns (FullIndexResponse);
    rpc registerIndexBarrel (IndexBarrelRequest) returns (ValidRegister);
    rpc getIndexBarrels (google.protobuf.Empty) returns (IndexBarrelInfo);
    rpc getStats(google.protobuf.Empty) returns (SystemStats);
    rpc updateIndexSize(IndexSizeUpdateRequest) returns (google.protobuf.Empty);
    rpc SendStats (SystemStats) returns (google.protobuf.Empty);
    rpc startSendingStatistics (WebServerInfo) returns (google.protobuf.Empty);
}

message PutNewRequest {
    string url = 1;
}

message AddToIndexRequestWords {
    string word = 1;
    string url = 2;
    bool from_multicast = 3;
}

message AddToIndexRequestLinks {
    string url = 1;
    string title = 2;
    string quote = 3;
    string link = 4;
    bool from_multicast = 5;
}

message SearchWordRequest {
    string words = 1;
}

message SearchBacklinksRequest {
    string url = 1;
}

message WordInfo {
    string url = 1;
    string title = 2;
    string quote = 3;
}

message SearchWordResponse {
    repeated WordInfo urls = 1;
}

message SearchBacklinksResponse {
    repeated string backlinks = 1;
}

message TakeNextResponse {
    string url = 1;
}

message FullIndexResponse {
    repeated IndexEntry palavras = 1;
    repeated IndexEntry urls = 2;
}

message IndexEntry {
    string param0 = 1;
    string param1 = 2;
}

message IndexBarrelRequest {
    string host = 1;
    string port = 2;
}

message ValidRegister{
    bool valid = 1;
}

message IndexBarrelInfo{
    repeated string indexInfos = 1;
}

message SystemStats {
  repeated string top_queries = 1;
  repeated BarrelStats barrels = 2;
}

message BarrelStats {
  string address = 1;
  int32 index_size_words = 2;
  int32 index_size_urls = 3;
  float avg_response_time = 4; 
}

message IndexSizeUpdateRequest {
  string host = 1;
  string port = 2;
  int32 lenIndexWords = 3;
  int32 lenIndexUrls = 4;
}

message WebServerInfo {
    string host = 1;
    string port = 2;
}